vim25/json/discriminator.go

```diff
@@ -184,19 +184,28 @@ func (d *decodeState) discriminatorGetValue() (reflect.Value, error) {
 			// Assign the type instance to the outer variable, t.
 			t = ti
 
-			switch t.Kind() {
-			case reflect.Map, reflect.Struct:
-				// If the type is a map or a struct then it is not necessary to
-				// continue walking over the current JSON object since it will be
-				// completely rescanned to decode its value into the discovered
-				// type.
-				dd.opcode = scanEndObject
-			default:
-				// Otherwise if the value offset has been discovered then it is
-				// safe to stop walking over the current JSON object as well.
+			// If type implements Unmarshaler we want to deserialize the
+			// value field into value of type t.
+			if supportsUnmarshaler(t) {
```

Comment:
@karaatanassov ,

I am struggling to understand why `supportsUnmarshaler` is invoked twice in `discriminatorGetValue`. Can you please add some code comments to both instances? I _think_:

* The first place that `supportsUnmarshaler` is doing so when the op indicates the value has already been read, thus this is to unmarshal the value now that the key has been detected.
* The second place is when the value op is discovered _after_ the key.

Is that the way of it?

Reply:
Yes. I think you are right.

If a type supports Unmarshaler then it's schema is always:

```
{
     "type": "<discriminator>",
     "value": <value>
}
```
In other words we do not care if the reflect type is struct/map or not.  We switch to the mode that is used for primitive types (not map or struct). 

I do not like how the "if ... switch" thing looks like and specifically copy of the code. I had a "never nester" pattern there with complex if condition and it was harder to read for me. So I changed it to this nested variant.
```
if supportsUnmarshaler(t) || !isMapOrStruct(t) {
```
I can revise it that form if it makes more sense to you.